docs/4.4/api-reference.md

```diff
@@ -5,63 +5,827 @@ description: The detailed guide to Teleport API
+  KubeUsers []string
+  // AppLabels is a map of labels used as part of the RBAC system.
+  AppLabels Labels
+  // ClusterLabels is a map of node labels (used to dynamically grant access to clusters).
+  ClusterLabels Labels
+}
+```
+
+**Labels**
+
+```go
+type Label map[string]utils.Strings
+```
+Labels are arbitrary key-value pairs that can be used to differentiate nodes, apps, or leaf clusters by key attributes. For example, `NodeLabels` might have the key `environment`, with its value set to `development`, `staging`, or `production`, according to the node's location.
+
+```go
+services.Labels{"environment": utils.Strings{"development", "staging"}}
+```
+
+Depending on which field you put these labels in, you can allow/deny access to any nodes, apps, or leaf clusters with the given labels. These labels can be very useful in systems where you need to carefully manage access across many clusters, e.g. if you are managing clusters for several outside groups.
+
+**Rules**
+
+The primary building blocks of a rule are its resources and verbs. The optional `Where` and `Actions` fields can be used for more advanced rules.
+
+```go
+type Rule struct {
+  // Resources is a list of resources
+  Resources []string
+  // Verbs is a list of verbs
+  Verbs []string
+  // Where specifies optional advanced matcher
+  Where string
+  // Actions specifies optional actions taken when this rule matches
+  Actions []string
+}
+
+```
+
+Here's an example of a rule describing `read only` verbs applied to the SSH `session` resource. Depending on if it's under `Allow` or `Deny`, it means "allow/deny users of this role the ability to read or list active SSH sessions".
+
+```go
+services.NewRule(
+  services.KindSession,
+  services.RO(), // helper function to get 'read only' verbs ("list" and "read")
+)
+```
+
+**Resources**
+
+These are all of the possible resource values, which can be found in the `services` package.
+
+```go
+KindUser              = "user"
+KindLicense           = "license"
+KindRole              = "role"
+KindAccessRequest     = "access_request"
+KindPluginData        = "plugin_data"
+KindOIDC              = "oidc"
+KindSAML              = "saml"
+KindGithub            = "github"
+KindOIDCRequest       = "oidc_request"
+KindSAMLRequest       = "saml_request"
+KindGithubRequest     = "github_request"
+KindWebSession        = "web_session"
+KindAuthServer        = "auth_server"
+KindProxy             = "proxy"
+KindNode              = "node"
+KindAppServer         = "app_server"
+KindToken             = "token"
+KindCertAuthority     = "cert_authority"
+KindOIDCConnector     = "oidc"
+KindSAMLConnector     = "saml"
+KindGithubConnector   = "github"
+KindClusterConfig     = "cluster_config"
+KindSemaphore         = "semaphore"
+KindClusterName       = "cluster_name"
+KindStaticTokens      = "static_tokens"
+KindTrustedCluster    = "trusted_cluster"
+KindIdentity          = "identity"
+KindKubeService       = "kube_service"
+```
```

Comment:
I wouldn't list these out like this unless we include documentation explaining how to use them.

Reply:
@russjones This list was a request from another review comment, because apparently a lot of users have trouble finding the values of resources to add to roles. Perhaps that is out of scope here though, and it definitely doesn't look pretty. I'll remove it for now.